src/client/TerritoryPatterns.ts

```diff
@@ -35,28 +30,38 @@ class PatternDecoder {
     if (version !== 1) {
       throw new Error("The pattern versions are different.");
     }
-    const tileWidth = (bytes[1] << 8) | bytes[2];
-    const tileHeight = (bytes[3] << 8) | bytes[4];
-    const scale = (bytes[5] << 8) | bytes[6];
 
-    const totalBits = tileWidth * tileHeight;
-    const totalBytes = Math.ceil(totalBits / 8);
-    const data = bytes.slice(7, 7 + totalBytes);
-    console.log("data", data);
+    this.tileWidth = (bytes[1] << 8) | bytes[2];
+    this.tileHeight = (bytes[3] << 8) | bytes[4];
+    this.scale = (bytes[5] << 8) | bytes[6];
+    this.dataStart = 7;
+    this.bytes = bytes;
+  }
 
-    return { data, tileWidth, tileHeight, scale };
+  getTileWidth(): number {
+    return this.tileWidth;
   }
-}
 
-export function initTerritoryPatterns(): void {
-  for (const [key, value] of Object.entries(territoryPatterns.patterns)) {
-    if (value.patternBase64) {
-      const decoded = PatternDecoder.decodeBase64Pattern(value.patternBase64);
-      value.patternData = decoded.data;
-      value.tileWidth = decoded.tileWidth;
-      value.tileHeight = decoded.tileHeight;
-      value.scale = decoded.scale;
-    }
+  getTileHeight(): number {
+    return this.tileHeight;
+  }
+
+  getScale(): number {
+    return this.scale;
+  }
+
+  isSet(x: number, y: number): boolean {
+    const px =
+      ((Math.floor(x / this.scale) % this.tileWidth) + this.tileWidth) %
+      this.tileWidth;
+    const py =
+      ((Math.floor(y / this.scale) % this.tileHeight) + this.tileHeight) %
+      this.tileHeight;
+    const idx = py * this.tileWidth + px;
+    const byteIndex = Math.floor(idx / 8);
```

Comment:
This is much more expensive than shift left:
```suggestion
    const byteIndex = idx << 3;
```